### PR TITLE
added SINDy library input option to WeakPDELibrary

### PR DESCRIPTION
**Pull request by znicolaou**
As a partial response to issue #351 and discussion #429 here, I'm adding an option to replace the `library_functions` and `function_names` input of `WeakPDELibrary` with `library`, which is an existing instance of `BaseFeatureLibrary`.  I've only tested so far on the first example in the example notebook 12, which reproduces the previous result:

```python
import numpy as np
import pysindy as ps
from scipy.integrate import solve_ivp
from pysindy.utils import lorenz

# Generate measurement data
dt = 0.002
t_train = np.arange(0, 10, dt)
t_train_span = (t_train[0], t_train[-1])
u0_train = [-8, 8, 27]
u_train = solve_ivp(
    lorenz, t_train_span, u0_train, t_eval=t_train
).y.T

# Define weak form ODE library
# defaults to derivative_order = 0 if not specified,
# and if spatial_grid is not specified, defaults to None,
# which allows weak form ODEs.
poly_lib = ps.PolynomialLibrary(
    degree=2,
    include_bias=False
)

ode_lib2 = ps.WeakPDELibrary(
    library=poly_lib,
    spatiotemporal_grid=t_train,
    is_uniform=True,
    K=100,
)


# Instantiate and fit the SINDy model with the integral of u_dot
model = ps.SINDy(feature_library=ode_lib2, optimizer)
model.fit(u_train)
model.print()
```

I'd also like to take some time to look through the PDE example notebooks to clean them up and make sure deprecated input has been removed, so I'm making this a draft for now. If anyone has other suggestions on modifications to the weak/pde input or functionality, this is a good chance. I may also look into #222 if it isn't a huge headache.

### CODE REVIEW

**Comment by Jacob-Stevens-Haas** [comment]
Thank you @znicolaou, this is a great feature!  I'll review the code today or tomorrow.  Do you have a simple test you could add for this?

I would defer #222 to a later PR, as it would help keep this PR simple.

**Comment by znicolaou** [comment]
Thanks, @Jacob-Stevens-Haas ! Yes, will revisit the tests and the example notebooks in the next day or two!

**Review by Jacob-Stevens-Haas — CHANGES_REQUESTED** [review]
This seems like such a great feature for a lot of reasons.  It expands the ability to use `WeakPDELibrary` for nontrivial problems, it avoids using special dunder attributes like `__code__`, and it simplifies the logic considerably.

Lets go ahead and deprecate the old behavior!  See the comment for a suggestion.

File: test/test_feature_library.py [hunk 1]
```diff
@@ -115,33 +115,59 @@ def test_pde_library_bad_parameters(params):
...
+        ),
+        dict(
+            spatiotemporal_grid=range(10),
+            H_xt=11,
+            library_functions=[lambda x: x, lambda x: x**2, lambda x: 0 * x],
+        ),
+        dict(
+            spatiotemporal_grid=range(10),
+            K=-1,
+            library_functions=[lambda x: x, lambda x: x**2, lambda x: 0 * x],
```
[the remark this answers is not in the document]
Comment (znicolaou) [reply test/test_feature_library.py:136]: Previously, the `library_functions=[lambda x: x, lambda x: x**2, lambda x: 0 * x] ` was specified in the `test_weak_pde_library_bad_parameters` in an odd way rather than in the `@pytest.mark.parametrize` part, and all the test used the same `library_functions`. Since I wanted to add the case with `library_functions=[]`, I restructured, which made more sense anyways.

File: pysindy/feature_library/weak_pde_library.py [hunk 2]
```diff
@@ -154,6 +158,7 @@ class WeakPDELibrary(BaseFeatureLibrary):
     def __init__(
         self,
         library_functions=[],
+        library=None,
```
[the remark this answers is not in the document]
Comment (znicolaou) [reply pysindy/feature_library/weak_pde_library.py:161]: Yes, that's how it should be done

File: pysindy/feature_library/weak_pde_library.py [hunk 3]
```diff
@@ -704,19 +722,23 @@ def get_feature_names(self, input_features=None):
...
+            lib_names = self.library.get_feature_names()
+        else:
+            for i, f in enumerate(self.functions):
+                for c in self._combinations(
+                    n_features, f.__code__.co_argcount, self.interaction_only
+                ):
+                    lib_names.append(
+                        self.function_names[i](*[input_features[j] for j in c])
+                    )
+        feature_names = feature_names + lib_names
```
[the remark this answers is not in the document]
Comment (znicolaou) [reply pysindy/feature_library/weak_pde_library.py:741]: I think that's a good idea. It'll take a bit to make the example notebooks reproduce the old behavior, which I think it necessary before we merge, but I think it's worth it.

File: pysindy/feature_library/weak_pde_library.py [hunk 4]
```diff
@@ -841,14 +861,20 @@ def transform(self, x_full):
...
+            else:
+                funcs = np.zeros((*x.shape[:-1], n_library_terms))
+                func_idx = 0
+
+                for f in self.functions:
+                    for c in self._combinations(
+                        n_features, f.__code__.co_argcount, self.interaction_only
+                    ):
+                        funcs[..., func_idx] = f(*[x[..., j] for j in c])
+                        func_idx += 1
```
[the remark this answers is not in the document]
Comment (znicolaou) [reply pysindy/feature_library/weak_pde_library.py:877]: ditto

**Comment by znicolaou** [comment]
Last point about deprecating--I think we should probably implement the same change in `PDELibrary` in addition, since it also used the same `library_functions` convention that is essentially repeating the `CustomLibrary`. Again, will take just a few small changes to the code but a bit of time to not break examples...

**Review by Jacob-Stevens-Haas — CHANGES_REQUESTED** [review]
Hey, I'm glad this spread to PDELibrary, but this adds a lot of new changes as well.  I'm OK with skipping deprecation and going right to removal of `library_functions` - after all, it does remove code and tests, and that's a good thing.  But why add a `TrimmedLibrary` class?

File: pysindy/feature_library/weak_pde_library.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (Jacob-Stevens-Haas) [reply pysindy/feature_library/weak_pde_library.py]: Wait, is this why you added `TrimmedLibrary`?

File: examples/1_feature_overview/example.py [hunk 5]
```diff
@@ -1,12 +1,12 @@
-# %% [markdown]
+#!/usr/bin/env python
```
[the remark this answers is not in the document]
Comment (znicolaou) [reply examples/1_feature_overview/example.py:1]: Ah, good to know about the publish_notebook script, I didn't keep track of that before. I'll do it in the future.

File: pysindy/feature_library/base.py [hunk 6]
```diff
@@ -506,6 +506,117 @@ def calc_trajectory(self, diff_method, x, t):
...
+            generated from applying the custom functions to the inputs.
+
+        """
+        check_is_fitted(self.library)
+
+        xp_full = []
+        for x in x_full:
+            xp = self.library.transform([x])[0]
+            xp = np.delete(xp, self.drop_inds, axis=xp.ax_coord)
+            xp = AxesArray(xp, comprehend_axes(xp))
```
[the remark this answers is not in the document]
Comment (znicolaou) [reply pysindy/feature_library/base.py:596]: Good question: I copied the format from the ConcatLibrary and made minimal modifications. I think it would work fine that way. As a side note, there's a lot of old code using AxesArray in inconsistent ways. It would be great to go through and ensure that the we don't rely on the old conventions of axis=-1 for ax_coord and axis=-2 for time, but I haven't had time to look through it.

File: pysindy/feature_library/pde_library.py [hunk 7]
```diff
@@ -103,12 +85,13 @@ class PDELibrary(BaseFeatureLibrary):
 
     def __init__(
         self,
-        library_functions=[],
+        function_library: BaseFeatureLibrary = PolynomialLibrary(
```
[the remark this answers is not in the document]
Comment (znicolaou) [reply pysindy/feature_library/pde_library.py:88]: Makes sense, thanks!

File: pysindy/optimizers/base.py [hunk 8]
```diff
@@ -93,7 +92,7 @@ def __init__(
         max_iter=20,
         normalize_columns=False,
         initial_guess=None,
-        copy_X=True,
+        copy_X=False,
```
[the remark this answers is not in the document]
Comment (znicolaou) [reply pysindy/optimizers/base.py:95]: The matrix being copied here is the output of the library.transform function after (passing through the SampleConcatter transform), which I don't think will ever be shared between two fits unless you hack it. For PDE data, this matrix can easily be more than 10GB in size, and copying it is very costly, especially when the copy does nothing. Looking at sklearn/linear_model/_base.py and `__init__` here, this `copy_X` is only used in the `_preprocess_data`. When `__fit_intercept=False` as is default here, `_preprocess_data` does no centering, no rescaling, so the copying seems to do nothing. Nevertheless, this was causing spikes in memory usage that can crash examples when memory is low. If there's a good reason to copy the data here, happy to note that you can set `copy_X=False` to conserve resources instead, but I'm pretty sure it's doing nothing.

File: pysindy/optimizers/stable_linear_sr3.py [hunk 9]
```diff
@@ -205,7 +207,7 @@ def _create_var_and_part_cost(
         y: np.ndarray,
         coef_sparse: np.ndarray,
         coef_neg_def: np.ndarray,
-    ) -> Tuple[cp.Variable, cp.Expression]:
+    ) -> Tuple["cp.Variable", "cp.Expression"]:
```
[the remark this answers is not in the document]
Comment (znicolaou) [reply pysindy/optimizers/stable_linear_sr3.py:210]: I just merged @akaptano's fix that he pushed in master. Can do this instead.

File: pysindy/feature_library/base.py [hunk 10]
```diff
@@ -506,6 +506,117 @@ def calc_trajectory(self, diff_method, x, t):
         return self.libraries[0].calc_trajectory(diff_method, x, t)
 
 
+class TrimmedLibrary(BaseFeatureLibrary):
```
[the remark this answers is not in the document]
Comment (znicolaou) [reply pysindy/feature_library/base.py:509]: When we made the `EnsembleOptimizer`, we lost some functionality in notebook 13, which I noticed when I went through the example notebooks now. It's also useful for implicit SINDy to be able to mask a feature library term if you want to use it in place of `x_dot`, as in `SINDyPI`. I added it here just to get the example notebook 13 running again, but it is useful to be able to construct a library then remove specific terms from it. Maybe it would be better to make another pull request, but I only had a week where I could focus on pysindy, so I snuck it in here. I have to turn to other work now.

File: pysindy/optimizers/base.py [hunk 11]
```diff
@@ -146,8 +145,8 @@ def fit(self, x_, y, sample_weight=None, **reduce_kws):
         x_ = AxesArray(np.asarray(x_), {"ax_sample": 0, "ax_coord": 1})
         y = AxesArray(np.asarray(y), {"ax_sample": 0, "ax_coord": 1})
         x_, y = drop_nan_samples(x_, y)
-        x_, y = check_X_y(x_, y, accept_sparse=[], y_numeric=True, multi_output=True)
 
+        # This calls check_X_y
```
[the remark this answers is not in the document]
Comment (znicolaou) [reply pysindy/optimizers/base.py:149]: Yeah, you're right, sorry about that.

**Comment by znicolaou** [comment]
I think this is done now, pending any other issues. Can take a look once more, but I have other work that I need to focus on now. If things look okay, happy to merge any time.

File: pysindy/optimizers/stable_linear_sr3.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (Jacob-Stevens-Haas) [reply pysindy/optimizers/stable_linear_sr3.py:210]: Ah, makes sense.  No worries - either way is fine, as SCA mostly reads quotes as forward references.

File: pysindy/feature_library/base.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (Jacob-Stevens-Haas) [reply pysindy/feature_library/base.py:509]: The quick fix, if you don't have time to focus on pysindy, is to either let example 13 be broken or, remove `TrimmedLibrary` from here and put it in notebook 13.  The enterprise-level fix is to remove it from this PR, add an issue for what was broken in 13 and why a `TrimmedLibrary` is a useful solution, and handle it separately with a new test.

Broadly speaking, I don't think we don't need to maintain the notebooks unless they implement a change for fast testing in #203 (see README.rst).  It's too much of a burden otherwise, since they take a huge amount of time and memory to run.  When someone runs into an error, they can make an issue and we can build a test, a fix, and if someone has the time, implement the maintainability fixes from the #203.

The broken notebooks serve as an example of what pysindy can do, even if they don't show the modern syntax.

File: pysindy/optimizers/base.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (Jacob-Stevens-Haas) [reply pysindy/optimizers/base.py:95]: I agree that it's helpful for `copy_X` to be false when fitting large data, but think the docstring is the place to note that option.  Changing the code would apply to all subclasses.  Making the default `True` would silently change the behavior of _any_ estimator that modifies the arguments to `_reduce` away from the convention not just in scikit-learn, but across all of PyData (e.g. numpy functions where `inplace=False` is default).  It's general practice that public functions should not mutate arguments.

Our sparse estimators have a public API, and I don't think using one outside of a `pysindy.SINDy` model is really a hack.  It's designed behavior.

File: pysindy/feature_library/base.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (Jacob-Stevens-Haas) [reply pysindy/feature_library/base.py:596]: Agree it would be nice, but it's probably not something anyone needs to put time into until it causes an error.

**Comment by Jacob-Stevens-Haas** [comment]
> I think this is done now, pending any other issues. Can take a look once more, but I have other work that I need to focus on now. If things look okay, happy to merge any time.

If you need to move onto other things, I can me to take this PR over the finish line if you'd like?

File: pysindy/feature_library/base.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (Jacob-Stevens-Haas) [reply pysindy/feature_library/base.py:509]: I agree that reproduciblility and supporting recent graduates is important.  But "How to keep one piece of software perfectly reproducible while working on another version" is a solved problem: add a requirements for each of the notebooks, and use a separate environment to satisfy those dependencies.  At the bare minimum, just put the version of pysindy that the experiments require at the top of the file: 1.7.4 or 1.7.5.  The alternative is never developing a new major version and never using a new major version of numpy, scikit-learn, matplotlib, etc, requiring future students and postdocs to maintain not just pysindy, but someone else's research project.

I didn't think you or Alan were that attached to the notebooks all being able to run, since when I started contributing a couple years ago, some of them were already broken.  Nevertheless, I do care about supporting your and Alan's research longevity, and that includes across major releases.  So I built #203 to make it easier for you and Alan to protect the functionality of your notebooks across major versions (in particular, see [the readme part](https://github.com/dynamicslab/pysindy#contributing-examples)).

There's probably a discussion to be had about examples vs full research papers as notebooks in the repo, as well as whether we should always should maintain backwards compatibility of a notebook across major versions even if they could have fast tests.  When does reading a bit of updated documentation or troubleshooting an error become negligible compared to reading the notebook & paper in the first place?

[the remark this answers is not in the document]
Comment (znicolaou) [reply pysindy/feature_library/base.py:509]: I'm trying to be respectful, but discussions like this are not a good use of resources. I can spend time explaining decisions to you that are important, but that's not really my job.  

You can open the notebooks, read the papers, understand what they are trying to do, and figure out why I made the changes if you want. You'll see that the features in the TrimmedLibrary are exactly those that were lost in notebook 13, but getting the notebook running is not the only reason to include it. I explained the use case with SINDyPI and ensemble SINDy already above, and I wanted to be able to test those things on research questions. 

I know you want me to make another pull request for it and you will weigh in on its value, but I don't want to waste a bunch of time writing about trivial changes and discussing them. If you disagree with my decisions, then I will hear out your rationale, but the costs of contributing here are outweighing the benefits for me. 

This repository is not an enterprise endeavor, and good scientific practices (particularly, not wasting time and computing resources) are at least as important as enterprise coding practices here in my opinion. We can have a difference of values, but we shouldn't let that become a barrier to collaboration.

**Comment by Jacob-Stevens-Haas** [comment]
With that PR and issue, are you ok if we close this, Zach?  (But let's keep the branch around so that we can pull in `TrimmedLibrary`)
